keps/sig-network/20190613-kube-proxy-component-config.md

```diff
@@ -0,0 +1,210 @@
+- To provide config structure, that is easier for users to understand and use.
+- To distinguish between instance local and shared settings.
+- To allow for the persistence of settings for different platforms (such as Linux and Windows) in a manner that reduces confusion and the possibility of an error.
+- To allow for easier introduction of new proxy backends.
+- To provide users with flexibility, especially with regards to the config source.
+
+### Non-Goals
+
+- To change or implement additional features in kube-proxy.
+- To deal with graduation of any other component of kube-proxy, other than its configuration.
+- To remove most or even all of the command line flags, that have corresponding component config options.
+
+## Proposal
+
+The idea is to conduct the process of graduation to beta in small steps in the span of at least one Kubernetes release cycle. This will be done by creating one or more alpha versions of the config with the last alpha version being renamed to v1beta1 after the community is happy with it.
+Each of the sub-sections below can result in a separate alpha version release, although it will be better for users to have no more than a couple of alpha versions past v1alpha1.
+The proposed process is similar to the already successfully used one for kubeadm.
+
+### Re-encapsulate mode specific options
+
+The current state of the config has proven that:
+- Some options are deemed as mode specific, but are in fact shared between all modes.
+- Some options are placed directly into KubeProxyConfiguration, but are in fact mode specific ones.
+- There are options that are shared between some (but not all) modes. Specific features of the underlying implementation are common and this happens only within the boundaries of the platform (iptables and ipvs modes for example).
+- Although legacy Linux and Windows user mode proxies are separate code bases, they have a common set of options.
+
+With that in mind, the following measures are proposed:
+- Mode specific structs are consolidated to not use fields from other mode specific structs.
+- Introduce a single combined legacy user mode proxy struct for both Linux and Windows backends.
+
+#### Example
+
+```yaml
+commonSetting1: ...
+commonSetting2: ...
+...
+modeA: ...
+modeB: ...
+modeC: ...
+```
+
+### Group platform specific fields together
+
+Platform specific fields need to be grouped together by platform name. This can be done by introducing structs to house them - one per platform (one for Linux, one for Windows, etc.). Then these structs will have an instance in the KubeProxyConfiguration.
+That can allow for users to specify settings for more than one platform per a YAML document, thus allowing for deduplication of platform independent shared fields. This can reduce the possibility of an error for users.
+
+#### Example
+
+```yaml
+commonSetting1: ...
+commonSetting2: ...
+...
+modeA: ... # modeA is platform independent
+platformX:
+  modeB: ...
+platformY:
+  modeC: ...
```

Comment:
Currently platform specific fields will be only the mode configs + ProxyMode (if we actually choose to have an explicit proxy mode). I like the idea about duplicating the platform independent modes on a per-platform basis.
So I think, that the proposed scheme is adequate, considering the kube-proxy config options ATM.

Reply:
> So I think, that the proposed scheme is adequate, considering the kube-proxy config options ATM.

Ok. Just to make sure we're on the same page, we'd have a pattern like this?

```
kind: KubeProxyConfiguration
clusterCIDR: ...
windows:
  someGeneralWindowsField: ...
  mode:
    kernel:
      networkName: ...
      sourceVIP: ...
      enableDSR: ...
```

I like that it makes clear via the platform type what modes you can use, gives us a place to put non-mode platform-specific stuff if it shows up, and maintains the 1:1 mode:platform pattern (we ought to document for developers that the same mode shouldn't be reused between platform types). 

From the example, it looks like we may also be planning something like this for legacy platform-independent modes:

```
kind: KubeProxyConfiguration
clusterCIDR: ...
legacyMode:
  userspace:
    ...
```

If so, I'd recommend instead duplicating the legacy platform-independent modes into platform-specific types when we refactor, so we can avoid carrying a legacy pattern in the new config. It will also be simpler to write validation if we don't have to cross-check platform-independent modes with the modes in a platform-specific type.

Reply:
This is a little abstract for me.  What are examples of "someGeneralWindowsField" or "someGeneralLinuxField" that are not better served by being de-de-duped under each mode?

Reply:
We don't have a good example of "someGeneralWindowsField" or "someGeneralLinuxField". In fact, I am a bit skeptical if we'll ever have such fields. I'd rather keep things as:
```
kind: KubeProxyConfiguration
clusterCIDR: ...
windows:
  kernel:
    networkName: ...
    sourceVIP: ...
    enableDSR: ...
```
and update the KEP accordingly.

Reply:
Yeah, agree with that. I also wasn't sure if such fields exist, or ever would.

Reply:
Though I think I still like `windows.mode.kernel` over `windows.kernel`, leaving some room for evolution just-in-case.